### PR TITLE
Use state transition bytecode from the database

### PR DESCRIPTION
**Pull request by xgreenx**
Related issue https://github.com/FuelLabs/fuel-core/issues/1754

Updated the upgradable executor to fetch the state transition bytecode from the database when the version doesn't match a native one. This change enables the WASM executor in the "production" build and requires a `wasm32-unknown-unknown` target.

The executor has a `VERSION` constant that must be updated when the behavior of the state transition function is changed. That constant is used to decide what executor to use(WASM or native).

## Checklist
- [x] New behavior is reflected in tests

### Before requesting review
- [x] I have reviewed the code myself

### CODE REVIEW

File: crates/services/upgradable-executor/src/executor.rs [hunk 1]
```diff
@@ -291,4 +404,407 @@ where
...
+
+                // Then
+                assert_eq!(Ok(()), result);
+            }
+        }
+
+        // The test verifies that `Executor::get_module` method caches the compiled WASM module.
+        // If it doesn't cache the modules, the test will fail with a timeout.
+        #[test]
+        #[ntest::timeout(40_000)]
```
Comment (Dentosal) [inline crates/services/upgradable-executor/src/executor.rs:791]: nit: Potentially flaky, especially in the CI. Cannot thing of an easy way to do this better, though.

File: crates/services/upgradable-executor/src/executor.rs [hunk 2]
```diff
@@ -291,4 +404,407 @@ where
...
+            let wrong_version = Executor::<Storage, DisabledRelayer>::VERSION + 1;
+            let block = valid_block(wrong_version);
+
+            // When
+            let result = executor
+                .execute_without_commit(ExecutionTypes::Validation(block))
+                .map(|_| ());
+
+            // Then
+            assert!(result.is_err());
```
Comment (Dentosal) [inline crates/services/upgradable-executor/src/executor.rs:636]: nit: use `.expect_err` for better error output

**Review by Dentosal — APPROVED** [review]
The logic itself looks good. I left a couple of nits on tests.

File: crates/services/upgradable-executor/src/executor.rs [hunk 3]
```diff
@@ -225,16 +307,48 @@ where
...
+        let block_version = block.state_transition_version();
+        if block_version == Self::VERSION {
+            match &self.execution_strategy {
+                ExecutionStrategy::Native => self.native_execute_inner(block, options),
+                ExecutionStrategy::Wasm { module } => {
+                    self.wasm_execute_inner(module, block, options)
+                }
+            }
+        } else {
+            let module = self.get_module(block_version)?;
```
Comment (Voxelot) [inline crates/services/upgradable-executor/src/executor.rs:319]: Do we have a way to ensure that the initial version of the executor is available in wasm? For example, if there have been several upgrades to the network and a node on the latest native version attempts to sync the network, will the bytecode for the initial executor version (before any upgrades occured) be available?

Comment (xgreenx) [reply crates/services/upgradable-executor/src/executor.rs:319]: No, we don't have a way. It should already be part of the snapshot/genesis block.

**Review by Dentosal — DISMISSED** [review]
LGTM. @Voxelot still has one open comment, so you might want to wait for him to review as well. Also needs conflict resolution.

File: crates/services/upgradable-executor/src/executor.rs (hunk 3 above)
Comment (xgreenx) [reply crates/services/upgradable-executor/src/executor.rs:319]: Automated the process of insertion of the Wasm bytecode into the `ChainConfig` https://github.com/FuelLabs/fuel-core/pull/1826

**Review by Dentosal — APPROVED** [review]

**Pull request merged**